Public/Src/Cache/ContentStore/Vfs/VfsTree.cs

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using BuildXL.Cache.ContentStore.Hashing;
+using BuildXL.Cache.ContentStore.Interfaces.FileSystem;
+using BuildXL.Cache.ContentStore.Interfaces.Sessions;
+using BuildXL.Cache.ContentStore.Utils;
+using BuildXL.Utilities.Collections;
+
+namespace BuildXL.Cache.ContentStore.Vfs
+{
+    /// <summary>
+    /// A store which virtualizes calls to an underlying content store (i.e. content will
+    /// be lazily materialized using the projected file system filter driver)
+    /// </summary>
+    internal class VfsTree
+    {
+        private readonly ConcurrentBigMap<string, VfsNode> _nodeMap = new ConcurrentBigMap<string, VfsNode>(keyComparer: StringComparer.OrdinalIgnoreCase);
+        private readonly VfsDirectoryNode _root;
+        private readonly VfsCasConfiguration _configuration;
+
+        public VfsTree(VfsCasConfiguration configuration)
+        {
+            _configuration = configuration;
+            _root = new VfsDirectoryNode(string.Empty, DateTime.UtcNow, null);
+            _nodeMap[string.Empty] = _root;
+        }
+
+        /// <summary>
+        /// Attempts to get the VFS node (directory or file) at the given VFS root relative path
+        /// </summary>
+        public bool TryGetNode(string relativePath, out VfsNode node)
+        {
+            return _nodeMap.TryGetValue(relativePath, out node);
+        }
+
+        /// <summary>
+        /// Adds a file node (and any needed parent directory nodes) at the VFS root relative path
+        /// </summary>
+        public VfsFileNode AddFileNode(string relativePath, VfsFilePlacementData data)
+        {
+            var timestamp = DateTime.UtcNow;

```

Comment:
I presume this is arbitrary and Vfs doesn't care if it's representative. Is that correct?

Reply:
Correct. It's mainly whether a file system consumer cares.

---
In reply to: [298427084](https://github.com/microsoft/BuildXL/pull/519#discussion_r298427084) [](ancestors = 298427084)